test/src/MNDiscovery/index.js

```diff
@@ -91,28 +96,54 @@ describe('MNDiscovery', async () => {
             const discovery = new MNDiscovery();
             sinon.spy(discovery.masternodeListProvider, 'getMNList');
 
-            var ips = [];
-            async function verifyRandomf(array) {
-                for (const item of array) {
-                    let randomMasternode = await discovery.getRandomMasternode();
-                    expect(masternodeIps).to.contain(randomMasternode.service.split(':')[0]);
-                    expect(randomMasternode.proRegTxHash).to.be.a('string');
-                    expect(randomMasternode.confirmedHash).to.be.a('string');
-                    expect(randomMasternode.votingAddress).to.be.a('string');
-                    expect(randomMasternode.pubKeyOperator).to.be.a('string');
-                    expect(randomMasternode.service).to.be.a('string');
-                    expect(randomMasternode.isValid).to.be.a('boolean');
-                    ips.push(randomMasternode.service.split(':')[0]);
-                }
-                expect(discovery.masternodeListProvider.getMNList.callCount).to.equal(array.length);
-                let uniqueIps = ips.filter(function (elem, pos) {
-                    return ips.indexOf(elem) == pos;
-                });
-                expect(uniqueIps.length > 1).to.be.true;
-                discovery.masternodeListProvider.getMNList.restore();
+            let ips = [];
+            const numRequests = 3;
+            for (let i = 0; i < numRequests; i++) {
+                let randomMasternode = await discovery.getRandomMasternode();
+                expect(masternodeIps).to.contain(randomMasternode.service.split(':')[0]);
+                expect(randomMasternode.proRegTxHash).to.be.a('string');
+                expect(randomMasternode.confirmedHash).to.be.a('string');
+                expect(randomMasternode.votingAddress).to.be.a('string');
+                expect(randomMasternode.pubKeyOperator).to.be.a('string');
+                expect(randomMasternode.service).to.be.a('string');
+                expect(randomMasternode.isValid).to.be.a('boolean');
+                ips.push(randomMasternode.service.split(':')[0]);
             }
+            expect(discovery.masternodeListProvider.getMNList.callCount).to.equal(numRequests);
+            let uniqueIps = ips.filter(function (elem, pos) {
+                return ips.indexOf(elem) == pos;
+            });
+            expect(uniqueIps.length > 1).to.be.true;
+            discovery.masternodeListProvider.getMNList.restore();
+        });
+
+        it('Should excludedIps from MN list', async () => {
+            const discovery = new MNDiscovery();
+            sinon.spy(discovery.masternodeListProvider, 'getMNList');
```

Comment:
Is it a unit test or functional? We will make a real request if `needsUpdate` is `true`.

Reply:
I would call them functional tests. I've added new test that verify that getRandomMasternode take into account excludedIP even when needsUpdate returns true